pkg/controllers/provisioning/provisioner.go

```diff
@@ -89,28 +76,27 @@ type Provisioner struct {
 }
 
 // Add a pod to the provisioner and block until it's processed. The caller
-// is responsible for verifying that the pod was scheduled correctly. In the
-// future, this may be expanded to include concepts such as retriable errors.
-func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) {
-	select {
-	case p.pods <- pod: // Block until pod is enqueued
-		<-p.wait
-	case <-p.running.Done(): // Leave if closed
-	}
+// is responsible for verifying that the pod was scheduled correctly.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) <-chan struct{} {
+	return p.batcher.Add(ctx, pod)
 }
 
 func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods, release when done
-	pods := p.batch(ctx)
-	defer func() {
-		for i := 0; i < len(pods); i++ {
-			p.wait <- struct{}{}
+	// Batch pods
+	logging.FromContext(ctx).Infof("Waiting for unschedulable pods")
+	items, window := p.batcher.Wait(ctx)
+	defer p.batcher.Flush()
```

Comment:
I think this defer Flush is still incorrect. 

When `Wait(ctx)` returns the items for the current batch, items can still be added to the batch before the function exits (so before the defer runs `Flush`).  The items added in-between the `Wait()` and the `Flush()` will be released when the function returns even though they were not in the items returned from this iteration of `Wait()`. 

Subsequent calls of `Wait` would return the items added in-between, but the chan returned from Add() would have already released the producer (pod controller).

Reply:
nvm, since the chan is unbuffered, Wait() has to be called before anything can be added.